### PR TITLE
Hobbies in Character Creation

### PR DESCRIPTION
**Pull request by Saelfen**
#### Summary
Features "Hobbies in Character Creation"

#### Purpose of change
Implement #45031
Players need flexibility in starting proficiency choices along with more specificity to character identity.

#### Describe the solution
Add Hobbies in character creation that act as multi-selectable professions which offer a package of proficiencies, skills, traits, and addictions.  Any number of hobbies can be chosen, giving more flexibility for character identity. This PR will be more sparse on the hobbies side, but there will be more to come.

image

On the skills side, hobbies provide additional experience instead of flat levels as a balancer against multiple hobby selections. The percentage and final level is reported in the description tab
image

#### Describe alternatives you've considered
Create a new hobbies object (and this is what I did to start), but it's essentially a direct copy of the existing professions object. So use that instead, and throw in a boolean to declare whether or not it's a hobby. There is no need to re-invent the wheel.

#### Testing
- Make sure tab order works with new Hobbies selection.
- Check Multi-Select works properly with increment/decrement points, skills, traits, final character description, and actually creating the character and starting the game. This includes loading proficiencies, skills, traits, and addictions.
- Make sure templates save and load properly.

#### Additional context
More hobbies to come in a second PR

### CODE REVIEW

**Comment by Broken27** [comment]
Add some Devices skill to the lockpicking hobby, it should have some.

.... and add a point cost to it, a real locksmith kit is pretty valuable. Certainly at least 1 point.

**Comment by AccountAlias** [comment]
Sure, I'll do that when I get some free time. Also, I'm extremely open to suggestions for hobbies. There's definitely a lot to add, and some to transfer from profession to hobbies

**Comment by Broken27** [comment]
Going by flavour text, camper, backpacker, Biker are all definetively hobbies.

Survivalist maybe too, although that one feels more professional.

The drug addiction starts should be hobbies too.

**Comment by Broken27** [comment]
Also why multiple hobbies is fine, i think we should have a limit of around 3 different ones tops. Or maybe even only 2.

(But unlimited in freeform so we can play a tweaker heroin addict crackhead chain smoker alcoholic).

**Comment by Saicchi** [comment]
There is no need to add a lot of hobbies now in this PR. Just add the minimum and you or others can add more later to make reviewing it easier.

**Comment by LaVeyanFiend** [comment]
Is there a way to turn "get direct levels from hobby" into "get XP from hobby so it has diminishing returns" as the linked issue wanted?

**Comment by AccountAlias** [comment]
> Also while multiple hobbies is fine, i think we should have a limit of around 3 different ones tops. Or maybe even only 2.

This makes sense, though I'm thinking the limit could also be 5. There's definitely discussion to be had for this

> There is no need to add a lot of hobbies now in this PR. Just add the minimum and you or others can add more later to make reviewing it easier.

Agreed, I won't go overboard on adding the hobbies then :)

> Is there a way to turn "get direct levels from hobby" into "get XP from hobby so it has diminishing returns" as the linked issue wanted?

Unfortunately, that's not so simple since we don't currently keep track of total cumulative exp for skills, and the experience is based on the formula 100 * 100 * (lvl + 1) * (lvl + 1). Scaling exp between the three different sources of skill (professions, hobbies, and skills) is, to me, complicated. Hence, why I left that part out.

File: src/newcharacter.cpp [hunk 1]
```diff
@@ -394,15 +395,21 @@ void avatar::randomize( const bool random_scenario, points_left &points, bool pl
...
 
     for( item &it : prof_items ) {
         if( it.has_flag( STATIC( flag_id( "WET" ) ) ) ) {
             it.active = true;
             it.item_counter = 450; // Give it some time to dry off
         }
+
         // TODO: debugmsg if food that isn't a seed is inedible
-        if( it.has_flag( json_flag_no_auto_equip ) ) {
+        else if( it.has_flag( json_flag_no_auto_equip ) ) {
```
Comment (Qrox) [inline src/newcharacter.cpp:412]: Why adding the else here? Items with the WET flag still needs to be handled according to the no-auto-equip/auto-wield flags.

Comment (AccountAlias) [reply src/newcharacter.cpp:412]: oh whoops, that was left over from when I tried to integrate hobbies into add_profession_items(). I ended up doing add_hobby_items() instead, and left the else in. Good catch!

**Comment by CirnoNeBaka** [comment]
Will items from multiple hobbies and professions add up? What if two hobbies provide the same item? I can see this possibly resulting in a multi-hobby character looking like a Christmas tree with multiple backpacks and hats :) 

What about scenarios like Naked and Afraid or Shower Victim? I think those need to override starting items from hobbies or their intended balance and flavor will be broken.

**Comment by AccountAlias** [comment]
> Will items from multiple hobbies and professions add up? What if two hobbies provide the same item? I can see this possibly resulting in a multi-hobby character looking like a Christmas tree with multiple backpacks and hats :)

Items from multiples hobbies do add up with professions. However, hobbies do not innately have your items worn/wielded. They act more like additional items you carry around as a result of your hobby. It is, however, a valid point on two hobbies providing the same item (albeit somewhat minor in impact). I'll take a look at that

> What about scenarios like Naked and Afraid or Shower Victim? I think those need to override starting items from hobbies or their intended balance and flavor will be broken.

Naked and Afraid and Shower Victim are professions, not scenarios. Adding cases for those specific scenarios where you're not supposed to have much is a lot of effort for little benefit (unless we hard-code to check for those specific professions, in which case its most likely a no). The easiest thing to do is to just drop your items and continue on

**Comment by Qrox** [comment]
For show victim, dropping the items (or putting them on a nearby table) is reasonable. For naked and afraid though, we should either remove all clothing items you get from hobbies, or make hobbies with starting clothing conflict with it. (You can't be both naked and cosplaying, for example.)

**Comment by CirnoNeBaka** [comment]
> Naked and Afraid and Shower Victim are professions, not scenarios. Adding cases for those specific scenarios where you're not supposed to have much is a lot of effort for little benefit (unless we hard-code to check for those specific professions, in which case its most likely a no). The easiest thing to do is to just drop your items and continue on

Another (less hard-coded) option I can think of would be to add a new optional json flag to professions that basically says "this profession doesn't get any items from hobbies". You can then check for this flag before calling `add_hobby_items()`.

**Comment by CirnoNeBaka** [comment]
Altough Shower Victim and Naked & Afraid should really be scenarios, not professions. Anyone from Otaku to Special Operator can become a shower victim during Cataclysm. But that's out of scope of this PR.

**Comment by Broken27** [comment]
We absolutely need a check to remove hobby items from scenarios where they don't make sense.

Otherwise we would get stuff like a convict starting with a full locksmith kit in prison because his hobby is lockpicking.

**Comment by AccountAlias** [comment]
> Otherwise we would get stuff like a convict starting with a full locksmith kit in prison because his hobby is lockpicking.

that's a great point

**Comment by I-am-Erk** [comment]
> 
> 
> > Also while multiple hobbies is fine, i think we should have a limit of around 3 different ones tops. Or maybe even only 2.
> 
> This makes sense, though I'm thinking the limit could also be 5. There's definitely discussion to be had for this
> 
> > There is no need to add a lot of hobbies now in this PR. Just add the minimum and you or others can add more later to make reviewing it easier.
> 
> Agreed, I won't go overboard on adding the hobbies then :)
> 
> > Is there a way to turn "get direct levels from hobby" into "get XP from hobby so it has diminishing returns" as the linked issue wanted?
> 
> Unfortunately, that's not so simple since we don't currently keep track of total cumulative exp for skills, and the experience is based on the formula 100 * 100 * (lvl + 1) * (lvl + 1). Scaling exp between the three different sources of skill (professions, hobbies, and skills) is, to me, complicated. Hence, why I left that part out.

Don't scale the exp based on the player level, that's the thing we want to avoid specifically to reduce stacking benefits. I would say that instead of giving you +1 level's worth of exp, professions and hobbies should give you 10000 * (profession skill bonus)^2 xp, so that it always assumes you are adding xp to a level 0 character, and taking just one profession won't change it but stacking sources won't give you exponentially increasing benefits.

**Comment by I-am-Erk** [comment]
> 
> 
> We absolutely need a check to remove hobby items from scenarios where they don't make sense.
> 
> Otherwise we would get stuff like a convict starting with a full locksmith kit in prison because his hobby is lockpicking.

I would make the argument that, at least to start, only your profession should give you items. Maybe later we could have a system like this, but it's a bit much for a first implementation PR. Alternatively we could later allow you to choose one 'loadout' based on the professions and hobbies you picked, so you don't get an item for each but rather decide which one we use.

File: tests/focus_test.cpp [hunk 2]
```diff
@@ -99,7 +99,7 @@ TEST_CASE( "focus" )
     SECTION( "time to level" ) {
         REQUIRE( you.get_skill_level( skill_id( "fabrication" ) ) == 0 );
         std::array<int, 11> expected_practice_times = {{
-                0, 173, 2137, 6303, 12137, 19637, 28803, 39637, 52137, 66303, 82137
+                0, 173, 2135, 6303, 12136, 19636, 28803, 39636, 52137, 66303, 82136
```
Comment (AccountAlias) [inline tests/focus_test.cpp:102]: change to setting `_exercise` from 0 to simply the difference between `_exercse` and `xp_to_level` leaves some carry-over exp, hence the minuscule reduction in turn count to level. This change is necessary from both a bug fix perspective and training skills based on experience obtained from hobbies

**Comment by LaVeyanFiend** [comment]
Are mods able to obsolete hobbies currently ex. primitive mod can get rid of firearm hobbies?

**Comment by AccountAlias** [comment]
Hobbies as per I-am-Erk will not deal with items anymore, so anything concerning items will not be an issue. Maybe in the future though. As for the actual question, I have no idea how mods work whatsoever, so I can't answer that question properly

**Comment by anothersimulacrum** [comment]
If we have a method of blacklisting professions, it can also blacklist hobbies AFAICT.

**Comment by AccountAlias** [comment]
No longer a draft, checks pass, and all the stuff I want in this initial implementation are in. More hobbies to come in a follow up PR

File: data/json/hobbies.json [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    "proficiencies": [ "prof_closures", "prof_leatherworking_basic" ]
+  },
+  {
+    "type": "profession",
+    "subtype": "hobby",
+    "id": "lockpicking",
+    "name": "Lockpicking",
+    "description": "Silently picking people's locks while they were away was a rush for you.  Will you still feel the same thrill now that no one is left to stop you?",
+    "points": 1,
+    "proficiencies": [ "prof_lockpicking" ]
```
Comment (I-am-Erk) [inline data/json/hobbies.json:73]:
```suggestion
    "proficiencies": [ "prof_lockpicking" ],
    "skills": [ { "level": 1, "name": "traps" } ]
```

File: data/json/hobbies.json [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  },
+  {
+    "type": "profession",
+    "subtype": "hobby",
+    "id": "parkour",
+    "name": "Parkour",
+    "description": "You've practiced parkour for many years, and made the world your playground.  It wouldn't be a lie to say that running is your life.  Which is good, because now that the end has come, you're running for your life.",
+    "points": 2,
+    "traits": [ "PARKOUR" ]
```
Comment (I-am-Erk) [inline data/json/hobbies.json:82]:
```suggestion
    "traits": [ "PARKOUR" ],
    "skills": [ { "level": 1, "name": "swimming" } ]
```

File: data/json/hobbies.json [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+  },
+  {
+    "type": "profession",
+    "subtype": "hobby",
+    "id": "camping",
```
Comment (I-am-Erk) [inline data/json/hobbies.json:87]: Might be worth refluffing this as a bit more wilderness survival-y given the skills

**Review by I-am-Erk — APPROVED** [review]
Some small tweaks to the JSON but this looks great to me.

**Comment by jiornojiovanni** [comment]
A "helicopter pilot" hobby would be cool, as a way to obtain the ability to use helicopters even with other professions.

**Comment by I-am-Erk** [comment]
A helicopter pilot hobby would be great. Please add it. We need many more hobbies now.

**Comment by faefux** [comment]
Calling addictions 'hobbies' is a bit... not right. I'm sorry if this is to 'social justicy' or whatever but must we be so disparaging about such serious afflictions in the way refer to them?

**Pull request merged**